File: leetcode/april2024/c++/22_open_the_lock.cpp
// CODE

class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> dead(deadends.begin(), deadends.end());
        unordered_set<string> vis;
        queue<string> q;
        
        q.push("0000");
        vis.insert("0000");

        if(dead.find("0000") != dead.end()) return -1;

        int res = 0;
        while(!q.empty()) {
            int sz = q.size();
            while(sz--) {
                string curr = q.front(); q.pop();

                if(curr == target) return res;
                if(dead.find(curr) != dead.end()) continue;

                for(int i = 0; i < 4; i++) {
                    char old = curr[i];
                    
                    // clockwise rotation
                    curr[i] = (old == '9' ? '0' : old + 1);
                    if(dead.find(curr) == dead.end() and vis.find(curr) == vis.end()) {
                        q.push(curr);
                        vis.insert(curr);
                    }

                    // anti-clockwise rotation
                    curr[i] = (old == '0' ? '9' : old - 1);
                    if(dead.find(curr) == dead.end() and vis.find(curr) == vis.end()) {
                        q.push(curr);
                        vis.insert(curr);
                    }

                    curr[i] = old;
                }
            }
            res++;
        }

        return -1;
    }
};

File: leetcode/april2024/python/22_open_the_lock.py
# CODE

from collections import deque

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        dead = set(deadends)
        vis = set()
        q = deque(["0000"])
        vis.add("0000")

        if "0000" in dead: return -1

        res = 0
        while q:
            sz = len(q)
            for _ in range(sz):
                curr = q.popleft()

                if curr == target: return res
                if curr in dead: continue

                for i in range(4):
                    old = int(curr[i])

                    # clockwise rotation
                    new_clockwise = (old + 1) % 10
                    new_curr_clockwise = curr[:i] + str(new_clockwise) + curr[i + 1:]
                    if new_curr_clockwise not in dead and new_curr_clockwise not in vis:
                        q.append(new_curr_clockwise)
                        vis.add(new_curr_clockwise)

                    # anti-clockwise rotation
                    new_anticlockwise = (old - 1) % 10
                    new_curr_anticlockwise = curr[:i] + str(new_anticlockwise) + curr[i + 1:]
                    if new_curr_anticlockwise not in dead and new_curr_anticlockwise not in vis:
                        q.append(new_curr_anticlockwise)
                        vis.add(new_curr_anticlockwise)

            res += 1

        return -1